client/consensus/epochs/src/lib.rs

```diff
@@ -798,6 +798,42 @@ where
 		});
 		self.epochs.insert((hash, number), persisted);
 	}
+
+	/// Revert to a specified block given its `hash` and `number`.
+	/// This removes all the epoch changes information that were announced by
+	/// all the given block descendents.
+	pub fn revert<D: IsDescendentOfBuilder<Hash>>(
+		&mut self,
+		descendent_of_builder: D,
+		hash: Hash,
+		number: Number,
+	) {
+		let is_descendent_of = descendent_of_builder.build_is_descendent_of(None);
+
+		let mut some_removed = false;
+		let mut predicate = |node_hash: &Hash, node_num: &Number, _: &PersistedEpochHeader<E>| {
+			if number >= *node_num &&
+				(is_descendent_of(node_hash, &hash).unwrap_or_default() || *node_hash == hash)
+			{
+				// Continue the search in this subtree.
+				(false, None)
+			} else if is_descendent_of(&hash, node_hash).unwrap_or_default() {
+				// Found a node to be removed.
+				some_removed = true;
+				(true, None)
+			} else if some_removed && *node_num < number {
+				// Backtrack detected, we can early stop the overall filtering operation.
+				(false, Some(true))
```

Comment:
IMO I'm not sure whether this is worth it, if we don't support stopping the search when backtracking the interface for `filter` would also be simpler, this optimization requires knowing that we're doing depth-first search and makes us depend on those implementation details.

Reply:
I'm not sure as well :-)
But I've found wasteful to continue analyzing the tree once that there is no chance to remove nothing else.

The depth-first strategy is documented for the `drain_filter` method, thus whoever writes the predicate can leverage this detail for some optimizations (as we did here).

But maybe you are right about keeping it simple and even more agnostic to the use case.